ls1intum/Artemis
# Convert build plan buttons to links

### Checklist
- [X] Client: I added multiple screenshots/screencasts of my UI changes(https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).

### Motivation and Context
Links to repositories of programming exercises are links and therefore one can open them directly in a new tab. The build plans, however, are still implemented as buttons that have the annoying property of only being left-clickable.

### Description
I removed the directives which were used to generate the links. I added a utils method that creates the links and added for all components a generation on startup which allows me to load the build plan link directly in the `href` attribute.
Since it's an external link, I added a `rel="noreferrer"`.

### Steps for Testing
This testing is only about checking whether all links are actually links (also those that look like buttons).
This includes
- Left-click transfers the user to a new tab
- Right-click opens the context menu
- Middle-click opens the page in the background
- Hovering over the link shows the link in the bottom corner
The term "make sure xxx is a link" refers to this.

**Exam**
1. Create an exam (no setup needed) with one programming exercise
2. Make sure that the two build plans (Template, Solution) in the exercise list are links and lead to the correct build plan

**Student**
1. Create one programming exercise with the build plan hidden and one with the build plan published
2. Participate as a student and make sure that the button link to the build plan is hidden for the first exercise
3. Participate as a student and make sure that the button to the build plan acts as a link and leads to the correct build plan

**Course**
1. Select one of the two previously created exercises for you to test
2. Make sure that the two build plans (Template, Solution) in the exercise list are links and lead to the correct build plan
3. Enter the exercise details and make sure that the two build plans (Template, Solution) are links and lead to the correct build plan
4. Navigate to the participations list and make sure that the build plan is a link and leads to the correct student build plan
5. Get back and navigate to the scores list. Make sure that the build plan button acts as a link and leads to the correct student build plan

### Review Progress
- Code Review
- Manual Tests
  - [x] Exam Exercise List 1
  - [x] Exam Exercise List 2
  - [x] Course ExerciseList 1
  - [x] Course ExerciseList 2
  - [x] Exercise Detail View 1
  - [x] Exercise Detail View 2
  - [x] Exercise Participation List 1
  - [x] Exercise Participation List 2
  - [x] Exercise Scores List 1
  - [x] Exercise Scores List 2
  - [x] Student Exercise View 1
  - [x] Student Exercise View 2

### Test Coverage
programming-exercise-group-cell.component.ts: 54.55%
programming-exercise-detail.component.ts: 44.14%
programming-exercise.component.ts: 72.53%
programming-exercise.utils.ts: 47.5%
exercise-scores.component.ts: 80.17%
participation.component.ts: 78.03%
course-exercise-details.component.ts: 75.86%
participation.utils.ts: 40%

### Screenshots
Publish Build Plan Setting on Exercise Creation
![Publish Build Plan Setting on Exercise Creation](https://user-images.githubusercontent.com/64264938/129967685-7153c0db-9f90-40b7-af30-b285f69b2c72.png)
Exam Exercise List
![Exam Exercise List](https://user-images.githubusercontent.com/64264938/129966021-6af153b8-59d0-4ee5-b4bf-33fcc9d7af76.png)
Course Exercise List
![Course Exercise List](https://user-images.githubusercontent.com/64264938/129966023-a879181d-6e10-431e-a4b5-b7e070bfe8d3.png)
Exercise Detail View (Course and Exam are the same)
![Exercise Detail View](https://user-images.githubusercontent.com/64264938/129966022-1bb0edcf-b18f-40b2-ba7c-6a2bab1bb0c6.png)
Exercise Participation List (Course and Exam are the same)
![Exercise Participation List](https://user-images.githubusercontent.com/64264938/129966019-54bb8406-c00a-479a-baec-2aafd82452ff.png)
Exercise Scores List (Course and Exam are the same)
![566c0f3298ea0396585a11620b3d44a4](https://user-images.githubusercontent.com/64264938/129966027-80030f73-7b5a-4715-a3f3-483a7a6356b8.png)
Student Exercise View with released Build Plan
![Student Exercise View with released Build Plan](https://user-images.githubusercontent.com/64264938/129966025-ebd2268e-f395-4041-8cad-8a355681cd35.png)
